third_party/CMakeLists.txt

```diff
@@ -57,11 +57,27 @@ elseif(WIN32)
 elseif(APPLE)
     target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_MACOS_MVK)
 elseif(UNIX)
-    # See whether X11 is available. If not, fall back to direct-to-display mode.
-    find_package(X11 QUIET)
-    if (X11_FOUND)
-        target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_XCB_KHR)
-    else()
+    # See whether XCB, Xlib or Wayland are available. If not, fall back to direct-to-display mode.
+    find_package(PkgConfig QUIET)
+    if (VKB_WSI_SELECTION STREQUAL XCB)
+        pkg_check_modules(XCB xcb QUIET)
+        if (XCB_FOUND)
+            target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_XCB_KHR)
+        endif()
+    endif()
+    if (VKB_WSI_SELECTION STREQUAL XLIB)
+        pkg_check_modules(X11 x11 QUIET)
+        if (X11_FOUND)
```

Comment:
Could we make direct-to-display part of VKB_WSI_SELECTION as well, rather than a fallback? Something like VKB_WSI_SELECTION=DIRECT would do it.

The code, as it stands, makes it difficult to switch from XCB, X11 or Wayland to direct-to-display, since the XCB_FOUND, X11_FOUND and WAYLAND_FOUND cache entries are still present.

This version worked for me, but you might want to add an error in the case that nothing is set in a final else case.
```
elseif(UNIX)
    # Choose WSI based on VKB_WSI_SELECTION
    find_package(PkgConfig QUIET)
    if (VKB_WSI_SELECTION STREQUAL XCB)
        pkg_check_modules(XCB xcb QUIET)
        if (XCB_FOUND)
            target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_XCB_KHR)
        endif()
    elseif (VKB_WSI_SELECTION STREQUAL XLIB)
        pkg_check_modules(X11 x11 QUIET)
        if (X11_FOUND)
            target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_XLIB_KHR)
        endif()
    elseif (VKB_WSI_SELECTION STREQUAL WAYLAND)
        pkg_check_modules(WAYLAND wayland-client QUIET)
        if (WAYLAND_FOUND)
            target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_WAYLAND_KHR)
        endif()
    elseif (VKB_WSI_SELECTION STREQUAL DIRECT)
        set(DIRECT_TO_DISPLAY TRUE)
        set(DIRECT_TO_DISPLAY TRUE PARENT_SCOPE)
        target_compile_definitions(vulkan INTERFACE VK_USE_PLATFORM_DISPLAY_KHR)
    endif()
endif()
```

You'd need to add DIRECT to `set(VKB_WSI_SELECTION "XCB" CACHE STRING "Select WSI target (XCB, XLIB, WAYLAND)")` as well of course.

We also have some issues building for Wayland with this at the moment. Those are under investigation.

Reply:
Direct-to-Display was selected as a fallback if X11 was not available but that's true, this VKB_WSI_SELECTION option can now be used to select Direct-to-Display.
I updated the third_party/CMakeLists.txt as suggested (but using D2D instead of DIRECT)
If I'm not wrong, the VKB_DIRECT_2_DISPLAY option can be removed (it's strange to me that the definition of VK_USE_PLATFORM_DISPLAY_KHR was not set for the build when using VKB_DIRECT_2_DISPLAY ...)

Reply:
Thanks for this.

I'm now wondering if all the QUIET flags for pkg_check_modules should actually be REQUIRED instead. There is no longer a fallback, so if you request a specific WSI it needs to be supported.

Reply:
Nice find! REQUIRED is now used instead of QUIET